src/replication.c

```diff
@@ -1380,14 +839,153 @@ handle_write_data_event(replica_t *replica)
 	return rc;
 }
 
+static void
+respond_with_error_for_all_outstanding_mgmt_ios(replica_t *r)
+{
+	MTX_LOCK(&r->r_mtx);
+	EMPTY_MGMT_Q_OF_REPLICA(r);
+	MTX_UNLOCK(&r->r_mtx);
+}
+
+static void
+inform_data_conn(replica_t *r)
+{
+	uint64_t num = 1;
+	r->disconnect_conn = 1;
+	if (write(r->mgmt_eventfd2, &num, sizeof (num)) != sizeof (num))
+		REPLICA_NOTICELOG("Failed to inform err to data_conn for replica(%p)\n", r);
+}
+
+static void
+free_replica(replica_t *r)
+{
+	/* TOOD: when to remove from rwaitq of spec */
+	pthread_mutex_destroy(&r->r_mtx);
+	pthread_cond_destroy(&r->r_cond);
+
+	free(r->mgmt_io_resp_hdr);
+	free(r->m_event1);
+	free(r->m_event2);
+
+	if (r->ip)
+		free(r->ip);
+	free(r);
+}
+
+void
+close_fd(int epollfd, int fd)
+{
+	if (epoll_ctl(epollfd, EPOLL_CTL_DEL, fd, NULL) == -1) {
+		perror("epoll_ctl");
+		return;
+	}
+	close(fd);
+}
+
+static void
+handle_mgmt_conn_error(replica_t *r, int sfd, struct epoll_event *events, int ev_count)
+{
+	int epollfd = r->epfd;
+	int mgmtfd, mgmt_eventfd1;
+	int i;
+	mgmt_event_t *mevent;
+	replica_t *r_ev;
+
+	MTX_LOCK(&r->spec->rq_mtx);
+	MTX_LOCK(&r->r_mtx);
+
+	r->conn_closed++;
+	if (r->conn_closed != 2) {
+		//ASSERT(r->conn_closed == 1);
+		/* case where error happened while sending HANDSHAKE or sending is successful but error from zvol_handshake */
+		if ((r->iofd == -1) || (r->mgmt_eventfd2 == -1)) {
+			TAILQ_FOREACH(r_ev, &(r->spec->rwaitq), r_waitnext) {
+				if (r_ev == r) {
+					TAILQ_REMOVE(&r->spec->rwaitq, r, r_waitnext);
+					r->conn_closed++;
+				}
+			}
+		} else {
```

Comment:
Informing replica thread to close data connection should be fine as long as mgmt_eventfd2 is not -1. For this, replica thread should set mgmt_eventfd2 to -1 with lock. Below is code change:
```diff --git a/src/data_conn.c b/src/data_conn.c
index 11b2f1e..8ae1b8f 100644
--- a/src/data_conn.c
+++ b/src/data_conn.c
@@ -192,7 +192,9 @@ handle_data_conn_error(replica_t *r)
 
        MTX_UNLOCK(&spec->rq_mtx);
 
+       MTX_LOCK(&r->r_mtx);
        if (epoll_ctl(r->epollfd, EPOLL_CTL_DEL, r->iofd, NULL) == -1) {
+               MTX_UNLOCK(&r->r_mtx);
                perror("epoll_ctl");
                return -1;
        }
@@ -200,16 +202,17 @@ handle_data_conn_error(replica_t *r)
        fd = r->iofd;
        r->iofd = -1;
 
-        if (r->io_resp_hdr)
-                free(r->io_resp_hdr);
-        if (r->ongoing_io_buf)
-                free(r->ongoing_io_buf);
-
        close(fd);
 
        data_eventfd = r->data_eventfd;
        r->data_eventfd = -1;
        close_fd(epollfd, data_eventfd);
+       MTX_UNLOCK(&r->r_mtx);
+
+       if (r->io_resp_hdr)
+               free(r->io_resp_hdr);
+       if (r->ongoing_io_buf)
+               free(r->ongoing_io_buf);
 
        respond_with_error_for_all_outstanding_ios(r);
 
@@ -225,6 +228,7 @@ handle_data_conn_error(replica_t *r)
         * and mgmt_conn thread, which are consumers of mgmt_eventfd2
         */
                pthread_cond_signal(&r->r_cond);
+       r->mgmt_eventfd2 = -1;
        MTX_UNLOCK(&r->r_mtx);
 
        /* replica might have got destroyed */
diff --git a/src/replication.c b/src/replication.c
index bf780c2..61d9461 100644
--- a/src/replication.c
+++ b/src/replication.c
@@ -897,7 +897,8 @@ handle_mgmt_conn_error(replica_t *r, int sfd, struct epoll_event *events, int ev
        r->conn_closed++;
        if (r->conn_closed != 2) {
                //ASSERT(r->conn_closed == 1);
-               /* case where error happened while sending HANDSHAKE or sending is successful but error from zvol_handshake */
+               /* case where error happened while sending HANDSHAKE or sending is successful but error from zvol_handshake or
+                * data connection is closed before this */
                if ((r->iofd == -1) || (r->mgmt_eventfd2 == -1)) {
                        TAILQ_FOREACH(r_ev, &(r->spec->rwaitq), r_waitnext) {
                                if (r_ev == r) {
@@ -905,7 +906,8 @@ handle_mgmt_conn_error(replica_t *r, int sfd, struct epoll_event *events, int ev
                                        r->conn_closed++;
                                }
                        }
-               } else {
+               }
+               if (r->mgmt_eventfd2 != -1) {
                        inform_data_conn(r);
                }
        } else {
```